### ISSUE TITLE
.NET X509Chain.Build() works on one Windows system but not another

### ISSUE THREAD

**Issue by samjpv** [dotnet/runtime#78254]
I have the following code to verify that a leaf certificate has been signed by a self-signed CA (sha256ecdsa):
```
using var chain = new X509Chain();
chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;  
chain.ChainPolicy.CustomTrustStore.Add(rootCert);  
chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;  
var builtChain = chain.Build(leafCert);  
if (!builtChain) return false;  
for (var i = 1; i < chain.ChainElements.Count; i++)  
{  
    if (chain.ChainElements[i].Certificate.Thumbprint != chain.ChainPolicy.CustomTrustStore[i - 1].Thumbprint)    
        return false;  
}  
return true;
```

When run locally on my Windows system, this check works, and builtChain = true. When I run this on either Linux or Windows on an Azure DevOps pipeline build agent, it fails with "certificate signature failure" as the chain element status for the leaf cert. I have also recreated this error running the analogous openssl command on my local machine:
`openssl verify -CAfile root.pem leaf.pem`
with output:
```
emailAddress = [REDACTED], CN = [REDACTED], O = [REDACTED], L = [REDACTED], ST = [REDACTED], C = US
error 7 at 0 depth lookup: certificate signature failure
error leaf.pem: verification failed
139747332990272:error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib:../crypto/asn1/a_verify.c:170:
```

So basically, running the .NET chain verification code succeeds, but verifying with openssl fails on my local system, and the errors I see from openssl on my local system are the same as I see on the build agent where it fails in the SDK. I have confirmed that both the build agent and my local machine use OpenSSL 1.1.1f  31 Mar 2020 and are building and running the solution on .NET 6. 

The public root cert is as follows:
```Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            [REDACTED]
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: CN = [REDACTED]
        Validity
            Not Before: Nov  3 22:35:41 2022 GMT
            Not After : Nov  2 22:35:41 2024 GMT
        Subject: CN = [REDACTED]
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:aa:64:20:9b:9b:31:66:62:14:f6:a5:fd:fd:90:
                    65:29:50:90:08:c8:b5:1e:c9:cf:f1:72:25:5f:a4:
                    3f:1c:68:e0:80:45:2a:11:c6:40:d4:a5:92:6d:05:
                    38:60:b8:36:66:56:ad:46:bb:6d:2c:31:4a:e8:0b:
                    f3:b8:9f:ca:e7
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                D9:2E:89:72:54:A2:6B:15:04:0A:4D:93:A8:BE:D9:CE:C2:1A:CA:AE
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:1
            X509v3 Key Usage:
                Certificate Sign, CRL Sign
    Signature Algorithm: ecdsa-with-SHA256
         30:45:02:20:0f:03:2c:0f:b7:06:7c:18:1d:bc:a2:6b:a5:92:
         db:9d:11:ac:a0:08:1b:f9:92:35:76:ab:71:23:fd:7a:bc:1f:
         02:21:00:95:4f:86:ae:38:29:c5:a8:b7:a5:6b:66:6e:44:3c:
         9c:fb:88:ac:4a:cb:89:ee:f6:47:c5:fc:34:83:7f:0d:fa
```

And the public leaf cert is as follows:
```Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: [REDACTED]
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: CN = [REDACTED]
        Validity
            Not Before: Nov 10 22:22:51 2022 GMT
            Not After : Nov 10 22:22:51 2023 GMT
        Subject: emailAddress = [REDACTED], CN = [REDACTED], O = [REDACTED], L = [REDACTED], ST = [REDACTED], C = US
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:92:ea:83:8b:4b:ee:e7:8b:3e:6f:df:b2:91:13:
                    bb:ac:fa:91:90:44:2f:17:1f:03:0c:d7:5e:d7:5d:
                    37:ca:81:b6:17:79:71:3e:aa:34:b0:3b:23:0f:00:
                    b5:ce:bf:f5:2b:37:12:50:ab:4c:e3:5b:17:0a:ea:
                    2e:f5:c0:49:b0
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Non Repudiation, Key Agreement
            X509v3 Basic Constraints: critical
                CA:FALSE, pathlen:0
            X509v3 Subject Key Identifier:
                1A:68:05:6C:91:F7:BE:0C:1A:9D:9C:C5:F5:F1:95:59:B5:A5:99:76
    Signature Algorithm: ecdsa-with-SHA256
         30:44:02:20:8c:23:32:07:24:03:40:88:42:31:50:5e:c3:ad:
         24:6c:ae:4c:1e:98:ed:64:ac:6a:83:48:51:f1:1d:4f:b4:58:
         02:20:e0:e8:50:58:b8:8b:b2:7d:66:a9:b9:f6:ea:73:45:d5:
         fc:e0:c2:e1:9d:22:7a:53:fe:a6:3d:d2:8f:ed:15:a9
```
         
For additional context, the leaf certificate has been signed with a sha256ecdsa signature that was manually converted from raw to asn1 format, HOWEVER, this signature is accepted when a device using the leaf certificate is registered to Azure IoT Hub. If the certificate were not valid, then registration using it would result in an error. 

Does anyone have thoughts on why this might be happening? Is it an issue with the SDK or something that I'm doing? Seems like something that @bartonjs might be able to help with.

**Comment by bartonjs** [comment]
> For additional context, the leaf certificate has been signed with a sha256ecdsa signature that was manually converted from raw to asn1 format

I'm not sure what this means.

---

If you can build a pair of certificates that you can share (no private keys should be needed) then we can look into this, but as described I don't think it's actionable.

**Comment by samjpv** [comment]
As far as the raw signature, the best resource on what I mean is probably here: https://stackoverflow.com/questions/63268843/invalid-signature-when-creating-a-certificate-using-bouncycastle-with-an-externa
But basically it had to be converted from
`R | S`
to
`0x30 | LEN(REMAININGDATA) | 0x02 | LEN (R) | R | 0x02 | LEN(S) | S`

Here are some test certs created the same way:

Root:
```
-----BEGIN CERTIFICATE-----
MIIBaTCCAQ6gAwIBAgIUK3e0/BUv6zcRlGxYJMeGLpa3pe0wCgYIKoZIzj0EAwIw
EjEQMA4GA1UEAwwHVGVzdCBDQTAeFw0yMjExMTEyMzI5NTBaFw0yNDExMTAyMzI5
NTBaMBIxEDAOBgNVBAMMB1Rlc3QgQ0EwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNC
AASTtMmJxrg72lT/qfC/L+j4L2ufZDKRqd1OV6j2lOQKMmsMAf31ePUqcuyVwUmS
4+D9AgSCMW3WULDetZ3KkQMHo0IwQDAdBgNVHQ4EFgQUv0wRldckyreKuu6AfplZ
SNjld90wEgYDVR0TAQH/BAgwBgEB/wIBATALBgNVHQ8EBAMCAQYwCgYIKoZIzj0E
AwIDSQAwRgIhALGDOO72T8zJHhmxRQIEOxk6vVHFAaZ61cVq+u8eqcE6AiEAuVrH
mbZ8JvNka94DA1lLU+1k7HdDsrX6jix+1UtXGck=
-----END CERTIFICATE-----

```

Leaf:
```
-----BEGIN CERTIFICATE-----
MIIBXDCCAQKgAwIBAgIIIpN5oK09Ro4wCgYIKoZIzj0EAwIwEjEQMA4GA1UEAwwH
VGVzdCBDQTAeFw0yMjExMTAyMzM1MzRaFw0yMzExMTAyMzM1MzRaMBIxEDAOBgNV
BAMTB1NBTVRFU1QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQW4nIY7RlDvrNy
B1ujP8nkc/hX9q449uG2IK+IS4sC9Z1L1B35za31AIOFa6xScn4gtNPLCPySoXkQ
ZPyKoenTo0IwQDAOBgNVHQ8BAf8EBAMCA8gwDwYDVR0TAQH/BAUwAwIBADAdBgNV
HQ4EFgQU2SPW0dE6YDvX3Dlxlo8XHDzvs9owCgYIKoZIzj0EAwIDSAAwRQIhABtg
xksUGz6x1SG9YM52VVN7PsySxPnypoTRHlp9iRajAiDckwvwXabB672am7u49Fkm
v59s1MP5QhODksPRN7J9CA==
-----END CERTIFICATE-----
```

**Comment by bartonjs** [comment]
Ah, that's quite useful to know.  Whatever produced that X.509-style ECDSA signature did it wrong.

Finding the signature block in the certificate and pretty printing it shows

```
$ xxd -r -p - | openssl asn1parse -i -inform der
30 45 02 21 00 1B 60 C6 4B 14 1B 3E B1 D5 21 BD
60 CE 76 55 53 7B 3E CC 92 C4 F9 F2 A6 84 D1 1E
5A 7D 89 16 A3 02 20 DC 93 0B F0 5D A6 C1 EB BD
9A 9B BB B8 F4 59 26 BF 9F 6C D4 C3 F9 42 13 83
92 C3 D1 37 B2 7D 08
    0:d=0  hl=2 l=  69 cons: SEQUENCE
    2:d=1  hl=2 l=  33 prim:  INTEGER           :BAD INTEGER:[001B60C64B141B3EB1D521BD60CE7655537B3ECC92C4F9F2A684D11E5A7D8916A3]
   37:d=1  hl=2 l=  32 prim:  INTEGER           :-236CF40FA2593E1442656444470BA6D94060932B3C06BDEC7C6D3C2EC84D82F8:[DC930BF05DA6C1EBBD9A9BBBB8F45926BF9F6CD4C3F942138392C3D137B27D08]
```

Whatever did the "add a 00 when necessary" got the check backwards.  It shouldn't be present in (r) (it has 9 zero-value-bits in a row, which is illegal), and it should be for (s) (because now it's negative).

Windows seems to have a high tolerance for user error here; OpenSSL apparently has low tolerance (and I agree with OpenSSL here).

If you create the certificates with the our builtin `CertificateRequest` class it'll do the right thing here; otherwise you need a better version of IEEE 1363P -&gt; X.509 DER encoding, like

https://source.dot.net/#System.Security.Cryptography/System/Security/Cryptography/AsymmetricAlgorithmHelpers.Der.cs,c91ec29287bbe91f

which needs

https://source.dot.net/#System.Security.Cryptography/src/libraries/Common/src/System/Security/Cryptography/KeyBlobHelpers.cs,de2ce52dbef74aba

**Comment by samjpv** [comment]
Ah, you're right. More of a whoever produced that signature, haha.. 

I made the mistake of padding on 1 for LSB instead of the MSB, rather than getting the check backwards altogether. I'm still a bit surprised that I was able to register on IoT Hub using the certificate, and that using Windows on the build agent didn't actually fix the issue. I suppose IoT Hub device provisioning might be forgiving, but I feel like OS might not be the issue on the SDK front. I'll have to look into it a bit further. Thanks @bartonjs!
